README.md
# STNEP
Transmission Network Expansion Planning (TNEP) using metaheuristics.
STNEP is a full non-convex AC formulation to solve the STNEP problem. It considers in an integrated fashion reactive power expansion, the (N-1) contingency criterion and operational costs. The formulation can be solved by different metaheuristics (IDE-PBILC, DE-PBILC, PSO and TLBO)


tspopf5.1_w64/CHANGES.txt
---------------------------
 Change history for TSPOPF
---------------------------

Version:    5.1

TSPOPF for MATPOWER
$Id: CHANGES.txt 2645 2015-03-11 20:53:01Z ray $
Copyright (c) 2007-2014 by Power System Engineering Research Center (PSERC)
See http://www.pserc.cornell.edu/tspopf/ for more info.


Changes since 5.0
-----------------

3/12/15
    - Released version 5.1 (no substantive changes from 5.0).

2/10/15
    - Replaced Inf*ones(...) with Inf(...) everywhere in tspopf_solver.m.

Changes since 5.0b1
-------------------

12/17/14
    - Released version 5.0 (no changes from 5.0b1).


Changes since 4.1
-----------------

7/1/14
    - Released version 5.0b1.

1/15/14
    - Added mpoption_info_tspopf() to define options for this
      optional package.

12/10/13
    - Updated for new MATPOWER options implementation based on options
      struct instead of options vector.

12/13/12
    - Minor modification to make it more robust in cases where some
      parameters are Inf (e.g. limits for dummy gen representing DC line).

5/3/12
    - Updated for compatibility with new @opf_model class based
      on the new @opt_model class, which supports indexed named sets of
      variables, constraints and costs.


Changes since 4.0
-----------------

11/11/11
    - Released version 4.1.

5/5/10
    - Added RETURN_RAW_DER option to control when OPF solver returns
      constraint, Jacobian and objective function gradient and Hessian
      information.

3/10/10
    - Added optional input arg to *ver() functions to trigger return
      of entire version struct with Name, Version, Release and Date
      just like MATLAB's ver() function.

2/23/10
    - For @opf_model, removed get_var_N(), get_lin_N() and get_nln_N()
      methods, replaced with single getN() method.


Changes since 1.0
-------------------

12/24/09
    - Released version 4.0 (skipped 2.x and 3.x to sync with MATPOWER).

11/4/09
    - Removed unnecessary 'return' statement at end of all M-files. If
      anything it should be an 'end' statement, but even that is
      optional, so we just let functions get terminated by the
      end-of-file or another function declaration.

10/23/09
    - Added build 17 for 64-bit Windows.

10/13/09
    - Added build 17 for 64-bit Linux.

9/22/09
    - Build 17, fixes for 64-bit compatibility only. Added build 17
      for 64-bit Intel Mac OS X.

6/4/09
    - Build 16a (PPC Mac only), fixed that incorrect mexopts file
      was used for Build 16 for PPC Mac.

4/21/09
    - Build 16
    - Fixed bug in bug in matlsolvSPMATRIX() where it claimed success
      even when the problem was so ill-conditioned that it returned
      NaN's in the solution vector.

3/31/09
    - Build 15
    - Updated date & version number (4.0).

3/30/09
    - Fixed bug in addSPMATRIX() where it could try to copy
      a matrix onto itself, resulting in a failed assertion.
    - Cleaned up printing of progress in IPM in OPT.c.

3/16/09
    - Since equal upper and lower bounds on a variable are not
      properly handled by the MEX file as an equality constraint,
      these bounds are relaxed and added as additional linear
      equality constraints, which are handled correctly by the MEX
      function.
    - Modified SPMATRIX2MATLABFILE() in SpMatrixLib_Helpers.c to
      write out the matrix to a more helpful MATLAB m-file.
      Original version created a script m-file that defined
      vectors r, c, x (row indices, col indices, values). New
      version creates a function m-file that returns the sparse
      matrix.

7/3/08
    - Fix to allow handling of user variables with any name, not
      just 'z'.
    - Added workaround to fix bug with pimul consistency with MINOPF.
      MINOPF adds a dummy row to pimul for piecewise linear costs
      only when the problem includes at least one piecewise linear
      cost. The TSPOPF MEX files always add this row, so we now
      strip it out when there are no piecewise linear costs.

5/22/08
    - Updated for major refactorization of OPF implementation in
      MATPOWER.

5/9/08
    - Fixed bug related to limits on z vars.
    - Fixed bug related to linear constraints unbounded below.

5/2/08
    - Move input arg processing to opf_args.m.


Version 1.0
-----------

9/25/07
    - Released version 1.0.
    - Build 14
    - Updated date & version number.

9/14/07
    - Build 13
    - Turned off DEBUG on Windows build to make sure it works on
      machines without the debug libraries.

9/12/07
    - Build 12
    - Fixed memory tromping bug in csrc2cscSPMATRIX().

9/11/07
    - Build 11
    - Fixed bug in SimPSERC.c:neutralStart() where all voltage
      angles, including non-zero reference angles, got set to zero.

8/14/07
    - Build 10
    - Fixed bugs in SCOPF_Cost1stDiff(), SCOPF_Cost2ndDiff() and
      SpMatVecProduct() that caused crashes when the N matrix in
      the cost had no non-zero elements.
    
8/10/07
    - Fixed bug in SCOPF_TestConvergence() caused by the assumption
      that the number of equality constraints equals nb*2.

8/9/07
    - Implement bounds on z variables in tspopf.m via general linear
      constraints, since they are not handled by the MEX files.
    - VERBOSE = 1 now prints version info.
    - Added some initialization of *_DEBUG variables for consistency.
    - Fixed pimul ordering in MEX file and removed corresponding
      workarounds in tspopf.m. Added multipliers on variable bounds.
    - Added explicit handling of linear equality constraints in the
      MEX files.

7/18/07
    - Build 09
    - Removed duplicate call to freeSPMATRIX(&ptrOPT->IPML2ndDiff)
      in freeOPT() in OPT.c.
    - Added zl and zh to MEX interface and call to SCOPF_Init(),
      not yet used by solvers.
    - Fixed bug where number of z variables was determined from N
      instead of A.

7/16/07
    - Build 08.
    - Added z to xr return variable.

7/6/07
    - Added ability in tspopf.m to specify bounds on user variables

STNEP/TestSystems.m
function [LineasIniciales, Costos]=TestSystems(sistema)
   
     %%-----  TEP Data  -----%%
     % Es un archivo para guardar los costos y la topologia inicial de cada
     % sistema
     
    switch sistema
%% Sistema garver 6 nodos 
        case {'garver_dina_cost_r','garver_sinred_sinreac_Edg','garver_sin_reac_CO','garver_sin_reac','garver_dina','garver_dina_cost','garver_sincomp_dina','garver_dina_CO','garver_dina_cost_CO','garver_sincomp_dina_CO','garver_sinrdpc_sincomp_dina'}
        LineasIniciales=[1 0 1 1 0 1 1 0 0 0 1 0 0 0 0];

        Costos=[40 38 60 20 68 20 40 31 30 59 20 48 63 30 61];
        
%% Sistema 24 nodos           
        case {'case24IEEE_dina', 'case24IEEE_dina_CO', 'case24IEEE_dina_Costo_R', 'case24IEEE_dina_Costo_R_CO', 'case24IEEE_dina_sinreactv', 'case24IEEE_dina_sinreactv_CO', 'case24IEEE_dina_sinrdpc_sinreactv'}
            LineasIniciales=[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 2 2 1 0 0 0 0 0 0 0];

            Costos=[  22 33 50 31 50 27 23 16 16 43 43 50 50 50 50 66 58 66 134 120 54 24 68 72 36 32 20 146 36 55 30 94 35 33 50 62 86 114 84];
%% Sistema 24 nodos  
        case {'case24IEEE_P_Glo_Edg','case24IEEE_P_Glo','case24IEEE_P_Glo_Costo_R','case24IEEE_P_Glo_CO','case24IEEE_P_Glo_Costo_R_CO'}
            LineasIniciales=[1	1	1	0	1	1	0	1	1	1	1	0	1	1	1	1	1	1	1	1	1	1	1	1	0	1	1	0	1	2	1	1	1	0	1	1	2	2	0	2	1];

            Costos=  [3 55 22 35 33 50 33 31 50	27	23	50	16	16	43	43	50	50	50	50	66	58	66	...
                    134	62	120	54	86	24	68	72	36	32	114	20	146	36	55	84	30	94];            

%% Sistema 87 nodos        
        case {'case87bus', 'case87bus_dc'}
        LineasIniciales=[2 0 0 1 0 0 0 0 1 0 0 0 0 0 0 1 1 2 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0 1 0 0 1 1 1 1 1 1 ... 
        1 3 2 1 3 1 1 0 1 0 0 2 2 2 4 2 3 2 2 1 0 1 3 0 2 2 0 0 3 2 1 2 2 1 2 1 0 0 0 0 0 0 0 0 0 0 0 ...
        1 1 2 0 0 0 0 1 1 0 0 0 2 0 1 0 4 0 0 0 2 0 4 1 0 1 1 1 1 2 0 0 0 1 2 0 1 1 0 2 3 0 2 3 1 0 0 ...
        0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0];

        Costos= [44056 48880 52230 31192 92253 60153 53253 21232 52230 58260 7510 26770 10020 20740 21232 26502 29852 8926 29182 27440 ...
        32130 48880 55580 21232 35212 7510 55580 56920 44860 21232 29048 54240 31460 23420 47540 53570 44190 51560 21232 7340 17390 ...
        6670 36284 35078 31326 7340 31594 30388 8926 21232 10690 26770 28780 35480 28780 28780 28780 20070 24760 13906 8926 50890 8926 27440 ...
        16720 21232 21678 21678 8926 40170 21232 5885 11165 6960 6435 12840 8926 12210 8926 9130 11935 7510 9900 12705 7510 29636 10120 ...
        8926 25500 12760 6710 8926 5335 25000 8926 9900 4510 4235 7510 8525 11660 7510 7150 6335 6215 12705 8085 9625 7315 4235 6105 7510 8926 5500 7510....
        4565 3465 31326 38160 10010 11660 4895 2090 12760 5335 3795 8926 32120 44860 21232 46870 25028 60940 25430 43520 21232 16050 21232 7510....
        45530 23420 7510 18060 55580 66300 21232 35480 18260 7510 35480 26100 21232 18060 21232 23240 18060 21232 125253 125253 80253 116253....
        149253 149253 149253 107253 125253 185253 131253 101253 128253 149253 101253 70953 173253 132753 137253 62253 98253 117753 125253 110253 23253 155253];

        case {'case87bus_sinrdptch', 'case87bus_AC', 'case87bus_AC_30percent_Renew_penetra', 'case87bus_AC_con_rdptch_sin_reactivos', 'case87bus_AC_wls'}

        LineasIniciales=[2 0 0 1 0 0 0 0 1 0 0 0 0 0 0 1 1 2 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0 1 0 0 1 1 1 1 1 1 ... 
        1 3 2 1 3 1 1 0 1 0 0 2 2 2 4 2 3 2 2 1 0 1 3 0 2 2 0 0 3 2 1 2 2 1 2 1 0 0 0 0 0 0 0 0 0 0 0 ...
        1 1 2 0 0 0 0 1 1 0 0 0 2 0 1 0 4 0 0 0 2 0 4 1 0 1 1 1 1 2 0 0 0 1 2 0 1 1 0 2 3 0 2 3 1 0 0 ...
        0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0];

        Costos= [44056 48880 52230 31192 92253 60153 53253 21232 52230 58260 7510 26770 10020 20740 21232 26502 29852 8926 29182 27440 ...
        32130 48880 55580 21232 35212 7510 55580 56920 44860 21232 29048 54240 31460 23420 47540 53570 44190 51560 21232 7340 17390 ...
        6670 36284 35078 31326 7340 31594 30388 8926 21232 10690 26770 28780 35480 28780 28780 28780 20070 24760 13906 8926 50890 8926 27440 ...
        16720 21232 21678 21678 8926 40170 21232 5885 11165 6960 6435 12840 8926 12210 8926 9130 11935 7510 9900 12705 7510 29636 10120 ...
        8926 25500 12760 6710 8926 5335 25000 8926 9900 4510 4235 7510 8525 11660 7510 7150 6335 6215 12705 8085 9625 7315 4235 6105 7510 8926 5500 7510....
        4565 3465 31326 38160 10010 11660 4895 2090 12760 5335 3795 8926 32120 44860 21232 46870 25028 60940 25430 43520 21232 16050 21232 7510....
        45530 23420 7510 18060 55580 66300 21232 35480 18260 7510 35480 26100 21232 18060 21232 23240 18060 21232 125253 125253 80253 116253....
        149253 149253 149253 107253 125253 185253 131253 101253 128253 149253 101253 70953 173253 132753 137253 62253 98253 117753 125253 110253 23253 155253];
%% Sistema 118 nodos    
        case {'case118_sin_comp_r','case118_cost_R_D','case118_cost_R_D_CO','case118_sin_comp','case118_sin_comp_CO'}
        LineasIniciales=[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1];
                     
        Costos= [18	          7.6	          1.8	         16.2	          9.7	          4.7...
            55	            6	           58	         12.4	         12.3	          4.4...
            11.1	           24	          6.1	         13.2	         12.7	         36.7...
            29.3	           15	          7.9	           27	          9.1	          8.9...
            17.6	          7.1	         15.3	         17.5	         23.9	          8.9...
            14.4	          6.9	         24.5	         15.4	           17	            7 ...
            91	          155	         23.4	            6	         17.3	         17.7...
            13.6	         18.7	         37.1	          2.3	          8.9	         21.3...
            6	          2.1	          6.8	         15.9	         25.2	           97	...
            10.9	          8.8	         27.5	         20.3	         36.8	         25.2	...
            16.2	         20.3	         19.1	         28.4	         11.3	         48.5	...
            48.5	         27.9	          9.1	         13.5	         20.6	         10.6	...
            24.5	         18.3	         43.4	         43.7	         12.7	          2.1	...
            3.4	         17.4	         20.1	         17.4	         12.9	         34.4	    ...
            37.7	         35.9	         32.4	         21.8	         22.5	            3	...
            10.1	          6.8	          6.9	           45	            6	          177	...
            54	         16.5	         16.5	         32.7	         17.6	          6.7	    ...
            15.2	           36	         41.7	         48.6	          6.7	         19.1	...
            61.7	          6.4	         29.4	           27	          8.2	         19.8	...
            21.2	         18.3	          7.3	         22.2	         15.2	           30	...
            2.8	          5.5	          8.7	         15.8	         12.7	           45	    ...
            6.7	         15.4	          6.6	         19.8	         22.2	         11.5	    ...
            18.5	         31.1	         15.3	           26	         12.8	         28.2	...
            17.9	           15	          9.1	         23.7	         19.1	         15.3	...
            23.7	         13.2	          7.8	         27.3	          9.5	         15.6	...
            16.8	         16.2	         30.9	         44.3	         10.4	          9.8	...
            15.9	         26.9	         14.6	         18.9	         10.1	         16.8	...
            9.5	         30.6	         23.8	         24.4	         34.4	          6.8	    ...
            9.8	         27.5	         12.7	         27.5	          6.5	         27.2	    ...
            13.7	         13.6	         11.5	          5.4	         30.5	           11	...
            13.3	          2.3	          0.9	           21	          8.7	          9.8];


%% Sistema 118 nodos costos anteriores    
        case {'case118_sin_comp_anter','case118_sin_comp_sin_redes_Edg'}
        LineasIniciales=[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...
                         1 1 1 1 1 1];
        Costos= [18000 7600 1800 16200 9700 4700 5500 6000 5800 12400 12300 4400 11100 24000 6100 13200 12700 36700 29300 15000 7900 27000 9100 8900 17600 7100 15300 17500 23900 8900 ...
                 14400 6900 24500 15400 17000 7000 9100 15500 23400 6000 17300 17700 13600 18700 37100 2300 8900 21300 6000 2100 6800 15900 25200 9700 10900 8800 27500 20300 36800 25200 ...
                 16200 20300 19100 28400 11300 48500 48500 27900 9100 13500 20600 10600 24500 18300 43400 43700 12700 2100 3400 17400 20100 17400 12900 34400 37700 35900 32400 21800 22500 3000 ...
                 10100 6800 6900 4500 6000 17700 5400 16500 16500 32700 17600 6700 15200 3600 41700 48600 6700 19100 61700 6400 29400 27000 8200 19800 21200 18300 7300 22200 15200 30000 ...
                 2800 5500 8700 15800 12700 4500 6700 15400 6600 19800 22200 11500 18500 31100 15300 26000 12800 28200 17900 15000 9100 23700 19100 15300 23700 13200 7800 27300 9500 15600 ...
                 16800 16200 30900 44300 10400 9800 15900 26900 14600 18900 10100 16800 9500 30600 23800 24400 34400 6800 9800 27500 12700 27500 6500 27200 13700 13600 11500 5400 30500 11000 ...
                 13300 2300 900 21000 8700 9800]./1000;


 %% Sistema 118 nodos    
        case {'case118P_Glo'}
        LineasIniciales=[1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0];
        Costos= [0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	4.06	3.248	4.06	5.075	2.842	4.466	3.248	3.654	6.293	5.278	4.06	3.248	4.06	5.075	2.842	4.466	3.248	3.654	3.248	5.278	1.048	1.048	3.248	1.048	2.842	1.066	3.248	1.054	3.248	1.048];            
             



%% Sistema 300 nodos
        
    case {'case300_sin_distr_cost_r','case300_sin_distr','case300_sin_distr_cost'}
        LineasIniciales=[1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	2	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	2	1	1	1	1	2	2	1	1 ...
                        1	1	2	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	2	1	1	1	1	1	1	1	1	1	2 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	2	2	1	1	1	2	1	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	2	1	1	1	1	1	1	1	1 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	2 ...
                        1	1	1	1	1	1	1	1	1	1	1	1	1	1	2	2	1	1	1	1	2 ...
                        1	1	1	1	1	2	1	1	1	2	1	2	1	1	1	1	1 ];
                    
         Costos=[9.1	15.5	23.4	6	17.3	37.7	33.6	48.7	57.1	2.3 ...
                 28.9	21.3	6	2.1	6.8	15.9	25.2	9.7	10.9	8.8	17.5	10.3 ...
                 56.8	25.2	26.2	10.3	15.1	28.4	11.3	13.5	38.5	17.9 ...
                 9.1	13.5	8	10.6	14.5	18.3	53.4	13.7	12.7	2.1	30.4 ...
                 17.4	20.1	17.4	12.9	14.4	17.7	15.9	12.4	41.8	12.5 ...
                 8.7	10.1	6.8	60.9	4.5	6	17.7	5.4	16.5	26.5	32.7	47.6 ...
                 6.7	15.2	3.6	11.7	18.6	6.7	19.1	11.7	6.4	19.4	17	8.2	19.8 ...
                 25.2	38.3	27.3	12.2	15.2	10	2.8	5.5	8.7	15.8	12.7	4.5	6.7 ...
                 15.4	36.6	29.8	12.2	11.5	18.5	11.1	15.3	16	12.8	18.2 ...
                 17.9	35	39.1	23.7	29.1	25.3	13.7	13.2	7.8	17.3	9.5	15.6 ...
                 16.8	16.2	10.9	14.3	10.4	9.8	15.9	16.9	14.6	18.9	10.1 ...
                 16.8	9.5	10.6	13.8	14.4	14.4	6.8	9.8	17.5	12.7	17.5	6.5 ...
                 17.2	13.7	13.6	11.5	5.4	20.5	11	13.3	2.3	9	21	8.7	9.8	15 ...
                 6.7	6.7	6.7	6.7	6.7	6.7	28	30	34	29.5	25	25.5	30.5	27	22.5 ...
                 29.5	27.5	30.5	23.5	23	6.7	27.1	30.6	27.9	29.5	29.5 ...
                 26.7	29.5	24.8	30.6	14.2	24.5	29	28.6	28	29.8	27.5 ...
                 29	30.2	36.7	29.9	39.1	26.8	27	6.7	6.7	24.7	23.8	13.8 ...
                 17.9	25.5	34.5	40	28.2	26.2	30.9	29.7	35.2	33.5	12.6 ...
                 6.6	17.1	13.1	17.7	11.4	13.3	17.5	14.5	14	16	25	21.2 ...
                 25.9	25	30.3	28.8	3	3.1	46	8	31.5	25.1	4.7	5.1	38.8	34.4 ...
                 43.7	10.7	14.3	16.3	10.7	6	19.5	8.6	12.7	11.8	16.5 ...
                 12.5	5.6	13.7	6	5	8.7	7	26.9	6	22.5	10.2	22.2	21.4 ...
                 20.5	8	22.8	45.7	17.8	24.4	19.8	20.4	6.7	27.8	35.9 ...
                 33.2	8.7	8	25	8	24.4	8	8	41.8	6	27.1	8	20.8	6.7	34.5 ...
                 6	6	8.4	10.3	11	9	11.2	8.4	7.5	8.2	9.3	6.7	6.7	8	6.7	6.7	6.7	21.1 ...
                 18.4	8.7	6.7	6.7	6.7	6.7	8	6.7	6.7	6.7	8	6.7	6.7	6.7	6	12.3	14.1 ...
                 6.7	21.2	20.9	6.7	6.7	6.7	20.3	21.3	6.7	6.7	6.7	6.7	6.7	6.7	20.7 ...
                 23.8	6.7	6	6	8	6	6	6	6	6	6	8	6.7	6	6	22.2	10.4 ...
                 6	6	6	8	6	8	8	8	6	8	6	6	6	6.7	8	6	8	6.7	6.7	6.7	6.7];        

    end
end

